### PR TITLE
Sink.ignore and foreach to log errors

### LINKED ISSUE #23501: Proposal: Sink.log or alternative run...() method to log completion/failure
We often recommend using `runWith(Sink.ignore)` or `runWith(Sink.foreach)` the issue with those is that those will not log the failure. And people are left wondering "why the stream does not complete" or "why the stream does not run".

If they'd add a `.log()` they'd realise "oh, the filename has a typo" (had this exact case now).

Perhaps such "terminating" one should log the failure? 
Perhaps this should be configurable? `akka.stream.verbose-error-sinks = on` or something?

I propose to make it easier for beginners to notice when they made a mistake that made a failure reach a Sink, the hello-world even exposes the need for this:

```
object TestServer extends App {
  implicit val system = ActorSystem()
  implicit val mat = ActorMaterializer()

FileIO.fromPath(new File("TYPO HERE").toPath)
    .via(Framing.delimiter(ByteString("\n"), maximumFrameLength = 1000))
    .map(_.utf8String)
    .map { line ⇒
      val entry = line.split(" ")
      val n = entry.head
      val time = entry.tail.head

      println(s"${n} ${Duration(time).toMicros}")
    } /*log ("progress")*/ runWith Sink.ignore
```

If I made a typo, I won't realise that I made a mistake. 

This of course is in-line with our error handling etc, however it's suprising to newcomers and we sometimes also get caught off guard by it (I just was, for a minute or so, but still).

Proposals (alternatives what we could do):

- Provide a Sink.log that by default would log completion and errors 
- Make our built-in "terminating, no outside communication" sinks like Sink.ignore log when the option `akka.stream.verbose-error-sinks = on` is on or something? This is a bit inconcistent, but could be pretty useful in debugging in any case, esp when an already live system has such issue and someone could not add the log() just like that.

### PR DESCRIPTION
**Pull request by richardimaoka**
Closes #23501
(Refs #23858, another approach which didn't seem to get consensus)
work in progress - documentation to be updated after #24522 is merged.

## description 
from https://github.com/akka/akka/pull/23858#issuecomment-364728824

To summarize the discussion so far, there seems to be two approaches:

1. As in#23858, use a config option to enable error logging in every stage:
 1.1 @patriknw advised this application-wide control is not appropriate https://github.com/akka/akka/pull/23858#discussion_r150556434 
 1.2. @ktoso is also on the fence about this. due to verbosity of error messages https://github.com/akka/akka/pull/23858#issue-149405110

2. The very original suggestion in https://github.com/akka/akka/issues/23501#issue-249239238 was actually to enable error logging in `Sink.ignore` and `Sink.foreach`, rather than error on every stage
 2.1 I submitted this PR #24523 for this approach

## test

[gist](https://gist.github.com/richardimaoka/dfc495b0494f0439ae1d958450cf404e)

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/scaladsl/Sink.scala [hunk 1]
```diff
@@ -235,7 +235,10 @@ object Sink {
   /**
    * A `Sink` that will consume the stream and discard the elements.
    */
-  def ignore: Sink[Any, Future[Done]] = fromGraph(GraphStages.IgnoreSink)
+  def ignore: Sink[Any, Future[Done]] =
+    Flow[Any].log("ignoreSink").toMat(ignoreWithoutLogging)(Keep.right)
```
Comment (richardimaoka) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Sink.scala]: As far as I checked, `def foreach` and `def onComplete` of the same `Sink` object (and Java API of `ignore`/`foreach`), which I already updated, were the only places calling this `ignore` other than the test code.

File: akka-stream/src/main/scala/akka/stream/scaladsl/Sink.scala [hunk 2]
```diff
@@ -244,7 +247,7 @@ object Sink {
    * the stream..
    */
   def foreach[T](f: T ⇒ Unit): Sink[T, Future[Done]] =
-    Flow[T].map(f).toMat(Sink.ignore)(Keep.right).named("foreachSink")
+    Flow[T].map(f).log("foreachSink").toMat(Sink.ignoreWithoutLogging)(Keep.right).named("foreachSink")
```
Comment (richardimaoka) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Sink.scala]: Is there a way to get this from attributes?

**Comment by richardimaoka** [comment]
Some tests, checking log output like `FlowLogSpec` failed. I will look into failure.

```
//FlowLogSpec
[info]   - must debug each element *** FAILED *** (2 seconds, 19 milliseconds)
[info]     java.lang.AssertionError: assertion failed: fishForMessage() found unexpected message Debug(akka.stream.Log(akka://FlowLogSpec/system/StreamSupervisor-0),class akka.stream.Materializer,[ignoreSink] Element: 2)
[info]     at scala.Predef$.assert(Predef.scala:219)
[info]     at akka.testkit.TestKitBase.recv$1(TestKit.scala:453)
```

**Review by patriknw — COMMENTED** [review]
I’m not sure this is the right approach. The sinks have a mat val for the purpose if you care about the result. It’s like if Future (or Future.foreach) would log each time they are completed with failure.

I’d rather make it easy to enable logging, eg withOnFailureLogging to make it convenient to add attribute that these and other stages can make use of.

Or with materializer setting.

Thanks for pushing the discussion forward.

**Comment by ktoso** [comment]
Hi @richardimaoka ! Thanks for keeping at it, as it definitely is a pretty important topic...
I agree with @patriknw, we can't do it ad hoc like this. I'm mostly tempted by a materializer setting nowadays, but still not sure how to best pull it off. It's a pretty tricky one!

**Comment by ktoso** [comment]
I'll close the PR but the topic is definitely still something we're interested in solving. Let's keep trying out things.

**Pull request closed without merging**